Marketplace: Add Jetpack Search free plan

epeicher opened the pull request:
#### Proposed Changes

* This PR adds the Jetpack Search free plan option when installing the `Jetpack Search` plugin. The new flow is described below. Described in text should be as described in the parent issue:

We should update the CTA logic with these two states:

- If neither version of the product is purchased/installed yet:
  - CTA should have the text “Start for Free”
  - Clicking it should redirect to checkout with `PRODUCT_JETPACK_SEARCH_FREE`
- If the free plan is purchased/installed:
  - CTA should have the text “Upgrade Jetpack Search”
  - Clicking it should redirect to checkout with `PRODUCT_JETPACK_SEARCH` or `PRODUCT_JETPACK_SEARCH _MONTHLY`)

```mermaid
flowchart TD 
	A[Start] --> B{Is Jetpack\nPaid installed?} 
	B --->|Yes| C[Plugin managed for you]
	B -->|No| D{Is Jetpack\nFree installed?}
	D -->|Yes| E[Upgrade Jetpack Search]
	D --> |No| F[Start for Free]
```
![CleanShot 2023-02-06 at 17 38 36@2x](https://user-images.githubusercontent.com/3519124/217030741-83c3ae7c-958d-446d-86f5-0a9737d84eae.png)

#### Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

##### Repeat the following steps on both a Simple site and an Atomic site
- Navigate to `/plugins/jetpack-search/:siteId`
- Make sure the CTA button reads `Start for Free`
- Click on the CTA button
- Make sure the system navigates to the checkout and the `Jetpack Search` is added to the Cart as a **free purchase**
- Proceed to checkout
- Check the checkout is successful
- Navigate to `/plugins/jetpack-search/:siteId`
- Make sure the CTA button reads `Upgrade Jetpack Search ` and the price is correctly displayed
- Click on the CTA button
- Make sure the system navigates to the checkout and the `Jetpack Search` is added to the Cart as a **paid products**. At this point, the user can select Monthly or Yearly
- Proceed to checkout
- Check the checkout is successful
- Navigate to `/plugins/jetpack-search/:siteId`
- Make sure there is no CTA button and a message is displayed instead indicating that the plugin is automatically managed for you.

#### Note: 
The following issue has been identified during development, and it will be done as a follow-up 
- https://github.com/Automattic/wp-calypso/issues/73009

#### Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [x] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [x] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [x] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
~- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?~
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Resolves #70333

epeicher committed:
Marketplace: Add Jetpack Search free plan

epeicher committed:
Handle null selectedSite and remove unused code

epeicher committed:
Remove not required Query component

The query will be performed by usePreinstalledPremiumPlugin hook

epeicher committed:
Do not render CTA actions when fetching

gcsecsey reviewed (changes requested):
With a simple site, it works great! However, on an Atomic site, when the free plan is installed and the CTA should display the upgrade, for me the screen is flickering and I can't click the CTA:
https://user-images.githubusercontent.com/11555574/217327016-10e3c2b9-2152-40ec-a153-4dfe60cf6b92.mp4

I've tried testing the flow again with a couple atomic site, both on Business and Ecommerce plans, I get the same result.

State | Simple site | Atomic site
--- | --- | ---
Note installed | <img width="1052" alt="CleanShot 2023-02-07 at 17 40 04@2x" src="https://user-images.githubusercontent.com/11555574/217322289-0a5bafaf-45e3-4ad0-948f-e873b62301a7.png"> | <img width="1054" alt="CleanShot 2023-02-07 at 17 43 36@2x" src="https://user-images.githubusercontent.com/11555574/217323232-7451a687-3636-4cbf-b215-df544ca098dc.png">
Free plan checkout | <img width="569" alt="CleanShot 2023-02-07 at 17 40 27@2x" src="https://user-images.githubusercontent.com/11555574/217322369-07178d28-3041-4f31-ba5b-b7dfda0b856c.png"> | <img width="564" alt="CleanShot 2023-02-07 at 17 44 01@2x" src="https://user-images.githubusercontent.com/11555574/217323354-0b4a7eb0-b7c8-42ab-8449-9150d32f007b.png">
Free plan installed | <img width="1060" alt="CleanShot 2023-02-07 at 17 41 02@2x" src="https://user-images.githubusercontent.com/11555574/217322466-ed283dcf-c7be-4c48-9efc-7cae5ccdc7f3.png"> | <img width="1059" alt="CleanShot 2023-02-07 at 17 46 27@2x" src="https://user-images.githubusercontent.com/11555574/217324910-1f0e4b3d-57a0-43ca-be83-908efbd05cb2.png">
Paid plan checkout | <img width="568" alt="CleanShot 2023-02-07 at 17 41 58@2x" src="https://user-images.githubusercontent.com/11555574/217322710-837804fa-5643-4e22-bf4f-3ca6a2e406c9.png"> | NA
Paid plan installed | <img width="1063" alt="CleanShot 2023-02-07 at 17 42 46@2x" src="https://user-images.githubusercontent.com/11555574/217322835-bce7d3f6-66d1-4699-988d-4dd7eb5f885c.png"> | NA

epeicher committed:
Remove loading flag that was causing an infinite loop

epeicher commented:
Good catch @gcsecsey! That was caused by [96cd353](https://github.com/Automattic/wp-calypso/pull/73006/commits/96cd3537f60686d5259385259fc3cbc9d35a1ca3) that was creating and infinite loop because the component was being destroyed and created again so[ this check ](https://github.com/Automattic/wp-calypso/blob/a07e7aa453bf47ea4fe023e7451aad5a97aba5bf/client/components/data/query-site-purchases/index.jsx#L19)was not doing its job properly. 

Anyways, the _loading_ placeholder is already on the main screen, so I don't think it is required here. I have removed the check and removed the `isLoading` flag that was causing the infinite loop toggling between `true` and `false`

gcsecsey reviewed (approved):
@epeicher thanks for swiftly investigating and fixing the issue! 🎉🙌

I've tested the changes again, and can confirm that the flow now works the same way for Atomic sites as well:

State | Atomic site
--- | ---
Free plan purchased | <img width="1064" alt="CleanShot 2023-02-08 at 13 29 26@2x" src="https://user-images.githubusercontent.com/11555574/217544273-5e4ff443-d903-4de8-ab6f-b9c32528a99d.png">
Paid plan checkout | <img width="563" alt="CleanShot 2023-02-08 at 13 29 40@2x" src="https://user-images.githubusercontent.com/11555574/217544301-2e7db9f9-27b5-441a-a873-f05084702341.png">
Paid plan purchased | <img width="1068" alt="CleanShot 2023-02-08 at 13 31 50@2x" src="https://user-images.githubusercontent.com/11555574/217544224-3b6e791b-60c7-44f1-8d1b-e0712be329ae.png">
